docs/design/resource-qos.md

```diff
@@ -0,0 +1,247 @@
+	name: foo
+		resources:
+	name: bar
+		resources:
+```
+
+Pods will not be killed if CPU guarantees cannot be met (for example if system tasks or daemons take up lots of CPU), they will be temporarily throttled.
+
+Memory is an incompressible resource and so let's discuss the semantics of memory management a bit.
+
+- *Best-Effort* pods will be treated as lowest priority. Processes in these pods are the first to get killed if the system runs out of memory.
+These containers can use any amount of free memory in the node though.
+
+- *Guaranteed* pods are considered top-priority and are guaranteed to not be killed until they exceed their limits, or if the system is under memory pressure and there are no lower priority containers that can be evicted.
+
+- *Burstable* pods have some form of minimal resource guarantee, but can use more resources when available.
+Under system memory pressure, these containers are more likely to be killed once they exceed their requests and no *Best-Effort* pods exist.
+
+### OOM Score configuration at the Nodes
+
+Pod OOM score configuration
+- Note that the OOM score of a process is 10 times the % of memory the process consumes, adjusted by OOM_SCORE_ADJ, barring exceptions (e.g. process is launched by root). Processes with higher OOM scores are killed.
+- The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will always be OOM killed before B.
+- The final OOM score of a process is also between 0 and 1000
+
+*Best-effort*
+	- Set OOM_SCORE_ADJ: 1000
+	- So processes in best-effort containers will have an OOM_SCORE of 1000
+
+*Guaranteed*
+	- Set OOM_SCORE_ADJ: -998
+	- So processes in guaranteed containers will have an OOM_SCORE of 0 or 1
+
+*Burstable*
+	- If total memory request > 99.8% of available memory, OOM_SCORE_ADJ: 2
+	- Otherwise, set OOM_SCORE_ADJ to 1000 - 10 * (% of memory requested)
+	- This ensures that the OOM_SCORE of burstable pod is > 1
+	- If memory request is `0`, OOM_SCORE_ADJ is set to `999`.
+	- So burstable pods will be killed if they conflict with guaranteed pods
+	- If a burstable pod uses less memory than requested, its OOM_SCORE < 1000
+	- So best-effort pods will be killed if they conflict with burstable pods using less than requested memory
+	- If a process in burstable pod's container uses more memory than what the container had requested, its OOM_SCORE will be 1000, if not its OOM_SCORE will be < 1000
+	- Assuming that a container typically has a single big process, if a burstable pod's container that uses more memory than requested conflicts with another burstable pod's container using less memory than requested, the former will be killed
+	- If burstable pod's containers with multiple processes conflict, then the formula for OOM scores is a heuristic, it will not ensure "Request and Limit" guarantees.
+
+
+*Pod infrastructure container*
+  - OOM_SCORE_ADJ: -998
+*Kubelet, Docker, Kube-Proxy*
```

Comment:
This is not true for kube-proxy anymore since it is running as daemonset already.